AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```diff
@@ -194,13 +232,69 @@ - (BOOL)setupAudioEngine {
      */
-    [_engine connect:_engine.inputNode to:_engine.mainMixerNode format:format];
+    [_recordEngine connect:_recordEngine.inputNode to:_recordEngine.mainMixerNode format:format];
 
-    _renderingContext->renderBlock = (__bridge void *)(_engine.manualRenderingBlock);
+    _capturingContext->renderBlock = (__bridge void *)(_recordEngine.manualRenderingBlock);
+
+    // Set the block to provide input data to engine
+    AVAudioInputNode *inputNode = _recordEngine.inputNode;
+    AudioBufferList *captureBufferList = &_captureBufferList;
+    success = [inputNode setManualRenderingInputPCMFormat:format
+                                               inputBlock: ^const AudioBufferList * _Nullable(AVAudioFrameCount inNumberOfFrames) {
+                                                   assert(inNumberOfFrames <= kMaximumFramesPerBuffer);
+                                                   return captureBufferList;
+                                               }];
+    if (!success) {
+        NSLog(@"Failed to set the manual rendering block");
+        return NO;
+    }
+
+    success = [_recordEngine startAndReturnError:&error];
+    if (!success) {
+        NSLog(@"Failed to start AVAudioEngine, error = %@", error);
+        return NO;
+    }
+
+    return YES;
+}
+
+- (BOOL)setupPlayoutAudioEngine {
+    NSAssert(_playoutEngine == nil, @"AVAudioEngine is already configured");
+
+    /*
+     * By default AVAudioEngine will render to/from the audio device, and automatically establish connections between
+     * nodes, e.g. inputNode -> effectNode -> outputNode.
+     */
+    _playoutEngine = [AVAudioEngine new];
+
+    // AVAudioEngine operates on the same format as the Core Audio output bus.
+    NSError *error = nil;
+    const AudioStreamBasicDescription asbd = [[[self class] activeFormat] streamDescription];
+    AVAudioFormat *format = [[AVAudioFormat alloc] initWithStreamDescription:&asbd];
+
+    // Switch to manual rendering mode
+    [_playoutEngine stop];
+    BOOL success = [_playoutEngine enableManualRenderingMode:AVAudioEngineManualRenderingModeRealtime
+                                                   format:format
+                                         maximumFrameCount:(uint32_t)kMaximumFramesPerBuffer
+                                                    error:&error];
+    if (!success) {
+        NSLog(@"Failed to setup manual rendering mode, error = %@", error);
+        return NO;
+    }
+
+    /*
+     * In manual rendering mode, AVAudioEngine won't receive audio from the microhpone. Instead, it will receive the
+     * audio data from the Video SDK and mix it in MainMixerNode. Here we connect the input node to the main mixer node.
+     * InputNode -> MainMixer -> OutputNode
+     */
+    [_playoutEngine connect:_playoutEngine.inputNode to:_playoutEngine.mainMixerNode format:format];
+
+    _renderingContext->renderBlock = (__bridge void *)(_playoutEngine.manualRenderingBlock);
```

Comment:
Is it worth it to mention the other source of input (the audio file) as well in the comment? Perhaps not if the player is added dynamically.

Reply:
Player is not added dynamically with this PR (it used to before). I can add an explanation here.